onnxruntime/core/optimizer/nhwc_transformer.cc

```diff
@@ -0,0 +1,366 @@
+         node.OpType() == "Relu" ||
+         node.OpType() == "Selu" ||
+         node.OpType() == "Sigmoid" ||
+         node.OpType() == "Softplus" ||
+         node.OpType() == "Softsign" ||
+         node.OpType() == "Tanh" ||
+         node.OpType() == "PRelu" ||
+         node.OpType() == "RandomNormal" ||
+         node.OpType() == "RandomUniform" ||
+         node.OpType() == "RandomNormalLike" ||
+         node.OpType() == "RandomUniformLike" ||
+         node.OpType() == "Multinomial")
+       layout = InheritLayout;
+  }
+
+  return layout;
+}
+
+void NhwcTransformerImpl::Transform(Node& node) {
+  long unsigned int node_idx = node.Index();
+
+  LOGS(logger, VERBOSE) << "NHWC " << node.OpType() << " "  << node.GetExecutionProviderType();
+
+  DataLayout required_layout = RequiredLayout(node);
+  bool bRequiresNHWC = (required_layout == NhwcLayout);
+  bool bNodeNHWC = false;
+
+  // create temporary container to allow inseration without alteration
+  std::vector<const Node::EdgeEnd*> inputEdges;
+  for (auto it = node.InputEdgesBegin(), end = node.InputEdgesEnd(); it != end; ++it)
+    inputEdges.push_back(&(*it));
+
+  for (std::vector<const Node::EdgeEnd*>::iterator it = inputEdges.begin(), end = inputEdges.end(); it != end; ++it) {
+    const Node::EdgeEnd* edge = *it;
+
+    bool bParentNHWC = true;
+
+    auto itParentLayout = nodes_layout.find(edge->GetNode().Index());
+    if (itParentLayout != nodes_layout.end())
+      bParentNHWC = itParentLayout->second;
+
+    switch (required_layout) {
+    case InheritLayout:
+        bNodeNHWC = bParentNHWC;
+        break;
+    case NhwcLayout:
+    case NchwLayout: {
+        if (bRequiresNHWC != bParentNHWC) {
+          NodeIndex permute_index = InsertPermuteParentNode(node, edge, bRequiresNHWC);
+          nodes_layout.insert(std::make_pair(permute_index, bRequiresNHWC));
+        }
+        bNodeNHWC = bRequiresNHWC;
+        break;
+      }
+    }
+  }
+
+  // first node
+  if (node.InputNodesBegin() == node.InputNodesEnd()) {
+    switch (required_layout) {
+    case InheritLayout:
+      bNodeNHWC = false;
+      break;
+    case NhwcLayout:
+    case NchwLayout:
+        if (bRequiresNHWC) {
+          NodeIndex permute_index = InsertPermuteParentNode(node, NULL, true);
+          nodes_layout.insert(std::make_pair(permute_index, true));
+        }
+        bNodeNHWC = bRequiresNHWC;
+        break;
+    }
+  }
+
+  // Replace specific ops
+  if (SuportsReplacementNHWC(node)) {
+    node_idx = ReplaceNode(node);
+    nodes_layout.insert(std::make_pair(node_idx, true));
+    bNodeNHWC = true;
+  }
+
+  nodes_layout.insert(std::make_pair(node_idx, bNodeNHWC));
+
+  Node& latest_node = *graph_.GetNode(node_idx);
+
+  if (latest_node.OutputNodesBegin() == latest_node.OutputNodesEnd() &&
+      bNodeNHWC) {
+    NodeIndex perute_index = InsertPermuteChildNode(latest_node, false);
+    nodes_layout.insert(std::make_pair(perute_index, false));
+  }
+}
```

Comment:
Is the intention of this to check if it's a graph output? If so it may be better to call Graph::GetNodeOutputsInGraphOutputs so you can handle a case where the node output is both a graph output and consumed by another node.

Reply:
No, the intention is to check if the node is a leaf since the insertion is different with regards to edges.